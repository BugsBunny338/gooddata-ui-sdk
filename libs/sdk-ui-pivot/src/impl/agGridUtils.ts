// (C) 2007-2020 GoodData Corporation
import { ICellRendererParams, ColDef, Column } from "@ag-grid-community/all-modules";
import { getMappingHeaderUri, IMappingHeader } from "@gooddata/sdk-ui";
import {
    DOT_PLACEHOLDER,
    FIELD_SEPARATOR,
    FIELD_SEPARATOR_PLACEHOLDER,
    ID_SEPARATOR,
    ID_SEPARATOR_PLACEHOLDER,
    ROW_SUBTOTAL,
    ROW_TOTAL,
    MEASURE_COLUMN,
} from "./agGridConst";
import { IGridHeader } from "./agGridTypes";
import escape = require("lodash/escape");
import {
    isTotalDescriptor,
    isMeasureGroupDescriptor,
    IExecutionResult,
    IAttributeDescriptor,
} from "@gooddata/sdk-backend-spi";
import { IDimension } from "@gooddata/sdk-model";
import invariant from "ts-invariant";

/*
 * Assorted utility functions used in our Pivot Table -> ag-grid integration.
 */

export const sanitizeField = (field: string) =>
    // Identifiers can not contain a dot character, because AGGrid cannot handle it.
    // Alternatively, we could handle it with a custom renderer (works in RowLoadingElement).
    field
        .replace(/\./g, DOT_PLACEHOLDER)
        .replace(new RegExp(FIELD_SEPARATOR, "g"), FIELD_SEPARATOR_PLACEHOLDER)
        .replace(new RegExp(ID_SEPARATOR, "g"), ID_SEPARATOR_PLACEHOLDER);

// returns [attributeId, attributeValueId]
// attributeValueId can be null if supplied with attribute uri instead of attribute value uri
export const getIdsFromUri = (uri: string, sanitize = true) => {
    const [, attributeId, , attributeValueId = null] = uri.match(/obj\/([^\/]*)(\/elements\?id=)?(.*)?$/);
    return [attributeId, attributeValueId].map((id: string | null) =>
        id && sanitize ? sanitizeField(id) : id,
    );
};

export const getParsedFields = (colId: string): string[][] => {
    // supported colIds are 'a_2009', 'a_2009_4-a_2071_12', 'a_2009_4-a_2071_12-m_3'
    return colId.split(FIELD_SEPARATOR).map((field: string) => field.split(ID_SEPARATOR));
};

export const colIdIsSimpleAttribute = (colId: string) => {
    const parsedFields = getParsedFields(colId);
    return parsedFields[0].length === 2 && parsedFields[0][0] === "a";
};

export const getRowNodeId = (item: any) => {
    return Object.keys(item.headerItemMap)
        .map((key) => {
            const mappingHeader: IMappingHeader = item.headerItemMap[key];

            if (isTotalDescriptor(mappingHeader)) {
                return `${key}${ID_SEPARATOR}${mappingHeader.totalHeaderItem.name}`;
            }

            const uri = getMappingHeaderUri(mappingHeader);
            const ids = getIdsFromUri(uri);
            return `${key}${ID_SEPARATOR}${ids[1]}`;
        })
        .join(FIELD_SEPARATOR);
};

export const getGridIndex = (position: number, gridDistance: number) => {
    return Math.floor(position / gridDistance);
};

export const cellRenderer = (params: ICellRendererParams) => {
    const isRowTotalOrSubtotal =
        params.data &&
        params.data.type &&
        (params.data.type === ROW_TOTAL || params.data.type === ROW_SUBTOTAL);

    const isActiveRowTotal =
        isRowTotalOrSubtotal && // short circuit for non row totals
        params.data &&
        params.data.rowTotalActiveMeasures &&
        params.data.rowTotalActiveMeasures.some((measureColId: string) =>
            params.colDef.field.endsWith(measureColId),
        );

    const formattedValue =
        isRowTotalOrSubtotal && !isActiveRowTotal && !params.value
            ? "" // inactive row total cells should be really empty (no "-") when they have no value (RAIL-1525)
            : escape(params.formatValue(params.value));
    const className = params.node.rowPinned === "top" ? "gd-sticky-header-value" : "s-value";
    return `<span class="${className}">${formattedValue || ""}</span>`;
};

export const getTreeLeaves = (tree: any, getChildren = (node: any) => node && node.children) => {
    const leaves = [];
    const nodes = Array.isArray(tree) ? [...tree] : [tree];
    let node;
    let children;
    while (
        // tslint:disable:no-conditional-assignment ban-comma-operator
        ((node = nodes.shift()),
        (children = getChildren(node)),
        (children && children.length) || (leaves.push(node) && nodes.length))
        // tslint:enable:no-conditional-assignment ban-comma-operator
    ) {
        if (children) {
            nodes.push(...children);
        }
    }
    return leaves;
};

export const indexOfTreeNode = (
    node: any,
    tree: any,
    matchNode = (nodeA: any, nodeB: any) => nodeA === nodeB,
    getChildren = (node: any) => (node && node.children) || [],
    indexes: number[] = [],
): number[] => {
    const nodes = Array.isArray(tree) ? [...tree] : [tree];
    for (let nodeIndex = 0; nodeIndex < nodes.length; nodeIndex++) {
        const currentNode = nodes[nodeIndex];
        // match current node
        if (matchNode(currentNode, node)) {
            return [...indexes, nodeIndex];
        }
        // check children
        const childrenMatchIndexes = indexOfTreeNode(node, getChildren(currentNode), matchNode, getChildren, [
            ...indexes,
            nodeIndex,
        ]);
        if (childrenMatchIndexes !== null) {
            return childrenMatchIndexes;
        }
    }
    return null;
};

export function getMeasureFormat(gridHeader: IGridHeader, result: IExecutionResult): string {
    const headers = result.dimensions[1].headers;
    const header = headers[headers.length - 1];

    if (!isMeasureGroupDescriptor(header)) {
        throw new Error(`Cannot get measure format from header ${Object.keys(header)}`);
    }

    const measureIndex = gridHeader.measureIndex;
    return header.measureGroupHeader.items[measureIndex].measureHeaderItem.format;
}

export function getSubtotalStyles(dimension: IDimension): string[] {
    if (!dimension || !dimension.totals) {
        return [];
    }

    let even = false;
    const subtotalStyles = dimension.itemIdentifiers.slice(1).map((attributeIdentifier) => {
        const hasSubtotal = dimension.totals.some(
            (total) => total.attributeIdentifier === attributeIdentifier,
        );

        if (hasSubtotal) {
            even = !even;
            return even ? "even" : "odd";
        }

        return null;
    });

    // Grand total (first) has no styles
    return [null, ...subtotalStyles];
}

export function getLastFieldType(fields: string[][]): string {
    const [lastFieldType] = fields[fields.length - 1];
    return lastFieldType;
}

export function getLastFieldId(fields: string[][]): string {
    const [, lastFieldId] = fields[fields.length - 1];
    return lastFieldId;
}

export function getAttributeLocators(fields: string[][], attributeHeaders: IAttributeDescriptor[]) {
    return fields.slice(0, -1).map((field: string[]) => {
        // first item is type which should be always 'a'
        const [, fieldId, fieldValueId] = field;
        const attributeHeaderMatch = attributeHeaders.find((attributeHeader: IAttributeDescriptor) => {
            return getIdsFromUri(attributeHeader.attributeHeader.formOf.uri)[0] === fieldId;
        });
        invariant(
            attributeHeaderMatch,
            `Could not find matching attribute header to field ${field.join(ID_SEPARATOR)}`,
        );
        return {
            attributeLocatorItem: {
                attributeIdentifier: attributeHeaderMatch.attributeHeader.localIdentifier,
                element: `${attributeHeaderMatch.attributeHeader.formOf.uri}/elements?id=${fieldValueId}`,
            },
        };
    });
}

export const getColumnIdentifierFromDef = (colDef: IGridHeader | ColDef): string => {
    // field should be always present, fallback to colId could happen for empty columns
    return colDef.field || colDef.colId;
};

export const getColumnIdentifier = (item: Column | IGridHeader | ColDef): string => {
    if (isColumn(item)) {
        return getColumnIdentifierFromDef(item.getColDef());
    }
    return getColumnIdentifierFromDef(item);
};

export function isColumn(item: Column | ColDef): item is Column {
    return !!(item as Column).getColDef;
}

export const isMeasureColumn = (item: Column | ColDef) => {
    if (isColumn(item)) {
        return item.getColDef().type === MEASURE_COLUMN;
    }
    return item.type === MEASURE_COLUMN;
};
