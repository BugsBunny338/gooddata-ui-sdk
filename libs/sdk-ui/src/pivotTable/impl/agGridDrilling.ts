// (C) 2007-2019 GoodData Corporation

import { getMappingHeaderUri } from "../../base/helpers/mappingHeader";
import { IMappingHeader } from "../../base/interfaces/MappingHeader";
import { getIdsFromUri } from "./agGridUtils";
import { COLUMN_ATTRIBUTE_COLUMN, MEASURE_COLUMN, ROW_ATTRIBUTE_COLUMN } from "./agGridConst";
import { ColDef } from "ag-grid-community";
import { IGridHeader } from "./agGridTypes";
import {
    IAttributeDescriptor,
    IDimensionItemDescriptor,
    IResultHeader,
    IResultMeasureHeader,
    isMeasureGroupDescriptor,
    isResultAttributeHeader,
    isResultMeasureHeader,
} from "@gooddata/sdk-backend-spi";
import get = require("lodash/get");

export const getDrillRowData = (leafColumnDefs: ColDef[], rowData: { [key: string]: any }) => {
    return leafColumnDefs.reduce((drillRow, colDef: ColDef) => {
        const { type } = colDef;
        // colDef without field is a utility column (e.g. top column label)
        if (colDef.field) {
            if (type === MEASURE_COLUMN) {
                return [...drillRow, rowData[colDef.field]];
            }
            const drillItem = get(rowData, ["headerItemMap", colDef.field]);
            if (drillItem && (type === COLUMN_ATTRIBUTE_COLUMN || type === ROW_ATTRIBUTE_COLUMN)) {
                const drillItemUri = getMappingHeaderUri(drillItem);
                return [
                    ...drillRow,
                    {
                        // Unlike fields, drilling data should not be sanitized, because it is not used in HTML properties
                        id: getIdsFromUri(drillItemUri, false)[1],
                        name: rowData[colDef.field],
                    },
                ];
            }
        }
        return drillRow;
    }, []);
};

export const getMeasureDrillItem = (
    responseHeaders: IDimensionItemDescriptor[],
    header: IResultMeasureHeader,
) => {
    const measureGroupHeader = responseHeaders.find(isMeasureGroupDescriptor);

    return get(measureGroupHeader, ["measureGroupHeader", "items", header.measureHeaderItem.order], null);
};

export const assignDrillItemsAndType = (
    header: IGridHeader,
    currentHeader: IResultHeader,
    responseHeaders: IDimensionItemDescriptor[],
    headerIndex: number,
    drillItems: IMappingHeader[],
) => {
    if (isResultAttributeHeader(currentHeader)) {
        header.type = COLUMN_ATTRIBUTE_COLUMN;
        // attribute value uri
        drillItems.push(currentHeader);
        // attribute uri and identifier
        const attributeDescriptor = responseHeaders[
            headerIndex % responseHeaders.length
        ] as IAttributeDescriptor;
        drillItems.push(attributeDescriptor);
        // This is where we could assign drillItems if we want to start drilling on column headers
        // It needs to have an empty array for some edge cases like column attributes without measures
    } else if (isResultMeasureHeader(currentHeader)) {
        // measure uri and identifier
        header.type = MEASURE_COLUMN;
        drillItems.push(getMeasureDrillItem(responseHeaders, currentHeader));
        header.drillItems = drillItems;
        header.measureIndex = currentHeader.measureHeaderItem.order;
    }
};
