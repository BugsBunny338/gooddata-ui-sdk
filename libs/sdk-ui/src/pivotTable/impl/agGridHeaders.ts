// (C) 2007-2019 GoodData Corporation

import { getMappingHeaderName } from "../../base/helpers/mappingHeader";
import { unwrap } from "../../base/helpers/utils";
import { IMappingHeader } from "../../base/interfaces/MappingHeader";
import { getIdsFromUri, getTreeLeaves } from "./agGridUtils";
import {
    COLUMN_GROUPING_DELIMITER,
    FIELD_SEPARATOR,
    ID_SEPARATOR,
    ROW_ATTRIBUTE_COLUMN,
} from "./agGridConst";
import { assignDrillItemsAndType } from "./agGridDrilling";
import { IColumnDefOptions, IGridAdapterOptions, IGridHeader, TableHeaders } from "./agGridTypes";
import {
    DataValue,
    DataViewFacade,
    IAttributeDescriptor,
    IDataView,
    IDimensionItemDescriptor,
    IMeasureDescriptor,
    IDimensionDescriptor,
    IResultHeader,
    isAttributeDescriptor,
    isMeasureGroupDescriptor,
    isResultAttributeHeader,
    isResultMeasureHeader,
    isResultTotalHeader,
} from "@gooddata/sdk-backend-spi";
import {
    IAttributeSortItem,
    IMeasureSortItem,
    isAttributeSort,
    isMeasureLocator,
    isMeasureSort,
} from "@gooddata/sdk-model";
import { ColDef } from "ag-grid-community";
import range = require("lodash/range");
import clone = require("lodash/clone");
import invariant = require("invariant");

/*
 * All code related to transforming headers from our backend to ag-grid specific data structures
 */

export const identifyHeader = (header: IResultHeader) => {
    if (isResultAttributeHeader(header)) {
        return `a${ID_SEPARATOR}${getIdsFromUri(header.attributeHeaderItem.uri).join(ID_SEPARATOR)}`;
    }
    if (isResultMeasureHeader(header)) {
        return `m${ID_SEPARATOR}${header.measureHeaderItem.order}`;
    }
    if (isResultTotalHeader(header)) {
        return `t${ID_SEPARATOR}${header.totalHeaderItem.type}`;
    }
    invariant(false, `Unknown header type: ${JSON.stringify(header)}`);
};

export const identifyResponseHeader = (header: IDimensionItemDescriptor) => {
    if (isAttributeDescriptor(header)) {
        // response headers have no value id
        return `a${ID_SEPARATOR}${getIdsFromUri(header.attributeHeader.uri)[0]}`;
    }
    if (isMeasureGroupDescriptor(header)) {
        // trying to identify a measure group would be ambiguous
        return null;
    }
    invariant(false, `Unknown response header type: ${JSON.stringify(header)}`);
};

export const headerToGrid = (header: IResultHeader, fieldPrefix = "") => {
    const internalHeader = unwrap(header);
    return {
        headerName: internalHeader.name,
        field: fieldPrefix + identifyHeader(header),
    };
};

export const shouldMergeHeaders = (
    resultHeaderDimension: IResultHeader[][],
    headerIndex: number,
    headerItemIndex: number,
): boolean => {
    for (let ancestorIndex = headerIndex; ancestorIndex >= 0; ancestorIndex--) {
        const currentAncestorHeader = resultHeaderDimension[ancestorIndex][headerItemIndex];
        const nextAncestorHeader = resultHeaderDimension[ancestorIndex][headerItemIndex + 1];
        if (
            !nextAncestorHeader ||
            identifyHeader(currentAncestorHeader) !== identifyHeader(nextAncestorHeader)
        ) {
            return false;
        }
    }
    return true;
};

export const mergeHeaderEndIndex = (
    resultHeaderDimension: IResultHeader[][],
    headerIndex: number,
    headerItemStartIndex: number,
): number => {
    const header = resultHeaderDimension[headerIndex];
    for (let headerItemIndex = headerItemStartIndex; headerItemIndex < header.length; headerItemIndex++) {
        if (!shouldMergeHeaders(resultHeaderDimension, headerIndex, headerItemIndex)) {
            return headerItemIndex;
        }
    }
    return headerItemStartIndex;
};

/*
 * getColumnHeaders transforms header items from matrix to tree hierarchy
 *Â for each span of identical headers in a row, the function is called recursively to assign child items
 */
export const getColumnHeaders = (
    resultHeaderDimension: IResultHeader[][],
    responseHeaders: IDimensionItemDescriptor[],
    columnDefOptions: IColumnDefOptions = {},
    headerIndex = 0,
    headerItemStartIndex = 0,
    headerValueEnd: number = undefined,
    fieldPrefix = "",
    parentDrillItems: IMappingHeader[] = [],
): IGridHeader[] => {
    if (!resultHeaderDimension.length) {
        return [];
    }

    const currentHeaders = resultHeaderDimension[headerIndex];
    const lastIndex = headerValueEnd !== undefined ? headerValueEnd : currentHeaders.length - 1;
    const hierarchy: IGridHeader[] = [];

    for (let headerItemIndex = headerItemStartIndex; headerItemIndex < lastIndex + 1; ) {
        const currentHeader = currentHeaders[headerItemIndex];
        const header: IGridHeader = {
            drillItems: [],
            ...headerToGrid(currentHeader, fieldPrefix),
            ...columnDefOptions,
        };
        const drillItems: IMappingHeader[] = clone(parentDrillItems);
        assignDrillItemsAndType(header, currentHeader, responseHeaders, headerIndex, drillItems);
        const headerItemEndIndex = mergeHeaderEndIndex(resultHeaderDimension, headerIndex, headerItemIndex);

        if (headerIndex !== resultHeaderDimension.length - 1) {
            header.children = getColumnHeaders(
                resultHeaderDimension,
                responseHeaders,
                columnDefOptions,
                headerIndex + 1,
                headerItemIndex,
                headerItemEndIndex,
                header.field + FIELD_SEPARATOR,
                drillItems,
            );
        }
        hierarchy.push(header);
        // We move the pointer manually to skip identical headers
        headerItemIndex = headerItemEndIndex + 1;
    }

    return hierarchy;
};

export const getRowHeaders = (
    rowDimensionHeaders: IAttributeDescriptor[],
    columnDefOptions: IColumnDefOptions,
    makeRowGroups: boolean,
): IGridHeader[] => {
    return rowDimensionHeaders.map((attributeHeader: IAttributeDescriptor) => {
        const rowGroupProps = makeRowGroups
            ? {
                  rowGroup: true,
                  hide: true,
              }
            : {};
        const field = identifyResponseHeader(attributeHeader);
        return {
            // The label should be attribute name (not attribute display form name)
            headerName: getMappingHeaderName(attributeHeader),
            type: ROW_ATTRIBUTE_COLUMN,
            // Row dimension must contain only attribute headers.
            field,
            drillItems: [attributeHeader],
            ...rowGroupProps,
            ...columnDefOptions,
        };
    });
};

export const getFields = (dataHeaders: IResultHeader[][]) => {
    return range((dataHeaders[0] || []).length).map((cellIndex: number) => {
        const fieldList = dataHeaders.map((header: IResultHeader[]) => identifyHeader(header[cellIndex]));
        return fieldList.join(FIELD_SEPARATOR);
    }) as string[];
};

// TODO: move this to data view facade / sanitize / make more generic
export const assortDimensionHeaders = (dimensions: IDimensionDescriptor[]) => {
    const dimensionHeaders: IDimensionItemDescriptor[] = dimensions.reduce(
        (headers: IDimensionItemDescriptor[], dimension: IDimensionDescriptor) => [
            ...headers,
            ...dimension.headers,
        ],
        [],
    );
    const attributeHeaders: IAttributeDescriptor[] = [];
    const measureHeaderItems: IMeasureDescriptor[] = [];
    dimensionHeaders.forEach((dimensionHeader: IDimensionItemDescriptor) => {
        if (isAttributeDescriptor(dimensionHeader)) {
            attributeHeaders.push(dimensionHeader);
        } else if (isMeasureGroupDescriptor(dimensionHeader)) {
            measureHeaderItems.push(...dimensionHeader.measureGroupHeader.items);
        }
    });
    return {
        attributeHeaders,
        measureHeaderItems,
    };
};

export const getMinimalRowData = (data: DataValue[][], rowHeaderItems: IResultHeader[][]) => {
    const numberOfRowHeaderItems = (rowHeaderItems[0] || []).length;

    return data.length > 0
        ? data
        : // if there are no measures only attributes
          // create array of [null] of length equal to the number of row dimension headerItems
          (Array(numberOfRowHeaderItems).fill([null]) as DataValue[][]);
};

const assignSorting = (colDef: ColDef, sortingMap: { [key: string]: string }): void => {
    const direction = sortingMap[colDef.field];
    if (direction) {
        colDef.sort = direction;
    }
};

/**
 * exported due to tests
 * @internal
 */
export const getAttributeSortItemFieldAndDirection = (
    sortItem: IAttributeSortItem,
    attributeHeaders: IAttributeDescriptor[],
): [string, string] => {
    const localIdentifier = sortItem.attributeSortItem.attributeIdentifier;

    const sortHeader = attributeHeaders.find(
        header => header.attributeHeader.localIdentifier === localIdentifier,
    );
    invariant(sortHeader, `Could not find sortHeader with localIdentifier ${localIdentifier}`);

    const field = identifyResponseHeader(sortHeader);
    return [field, sortItem.attributeSortItem.direction];
};

/**
 * exported due to tests
 * @internal
 */
export const getMeasureSortItemFieldAndDirection = (
    sortItem: IMeasureSortItem,
    measureHeaderItems: IMeasureDescriptor[],
): [string, string] => {
    const keys: string[] = [];
    sortItem.measureSortItem.locators.forEach(locator => {
        if (isMeasureLocator(locator)) {
            const measureSortHeaderIndex = measureHeaderItems.findIndex(
                measureHeaderItem =>
                    measureHeaderItem.measureHeaderItem.localIdentifier ===
                    locator.measureLocatorItem.measureIdentifier,
            );
            keys.push(`m${ID_SEPARATOR}${measureSortHeaderIndex}`);
        } else {
            const key = `a${ID_SEPARATOR}${getIdsFromUri(locator.attributeLocatorItem.element).join(
                ID_SEPARATOR,
            )}`;
            keys.push(key);
        }
    });
    const field = keys.join(FIELD_SEPARATOR);
    const direction = sortItem.measureSortItem.direction;
    return [field, direction];
};

export function createTableHeaders(dataView: IDataView, options: IGridAdapterOptions = {}): TableHeaders {
    const dv = new DataViewFacade(dataView);
    const dimensions = dv.dimensions();
    const headerItems = dv.headerItems();
    const { columnDefOptions, makeRowGroups = false } = options;

    const sorting = dv.definition.sortBy;
    const sortingMap = {};
    const { attributeHeaders, measureHeaderItems } = assortDimensionHeaders(dimensions);
    sorting.forEach(sortItem => {
        if (isAttributeSort(sortItem)) {
            const [field, direction] = getAttributeSortItemFieldAndDirection(sortItem, attributeHeaders);
            sortingMap[field] = direction;
        }
        if (isMeasureSort(sortItem)) {
            const [field, direction] = getMeasureSortItemFieldAndDirection(sortItem, measureHeaderItems);
            sortingMap[field] = direction;
        }
    });

    const columnAttributeHeaderCount = dimensions[1].headers.filter(
        (header: IDimensionItemDescriptor) => !!(header as IAttributeDescriptor).attributeHeader,
    ).length;

    const columnHeaders: IGridHeader[] = getColumnHeaders(
        headerItems[1],
        dimensions[1].headers,
        columnDefOptions,
    );

    const groupColumnHeaders: IGridHeader[] =
        columnAttributeHeaderCount > 0
            ? [
                  {
                      headerName: dimensions[1].headers
                          .filter(header => isAttributeDescriptor(header))
                          .map((header: IAttributeDescriptor) => {
                              return getMappingHeaderName(header);
                          })
                          .filter((item: string) => item !== null)
                          .join(COLUMN_GROUPING_DELIMITER),
                      field: "columnGroupLabel",
                      children: columnHeaders,
                      drillItems: [],
                  },
              ]
            : columnHeaders;

    const rowHeaders: IGridHeader[] =
        // There are supposed to be only attribute headers on the first dimension
        getRowHeaders(dimensions[0].headers as IAttributeDescriptor[], columnDefOptions, makeRowGroups);

    const allHeaders: IGridHeader[] = [...rowHeaders, ...groupColumnHeaders].map((column, index) => {
        if (column.children) {
            getTreeLeaves(column).forEach((leafColumn, leafColumnIndex) => {
                leafColumn.index = index + leafColumnIndex;
                assignSorting(leafColumn, sortingMap);
            });
        }
        column.index = index;
        assignSorting(column, sortingMap);
        return column;
    });

    const colFields: string[] = getFields(headerItems[1]);
    const rowFields: string[] = rowHeaders.map(header => header.field);

    const leafColumnDefs = getTreeLeaves(allHeaders);
    if (leafColumnDefs[0]) {
        leafColumnDefs[0].cellRenderer = "loadingRenderer";
    }

    return {
        rowHeaders,
        colHeaders: groupColumnHeaders,
        allHeaders,
        rowFields,
        colFields,
    };
}
