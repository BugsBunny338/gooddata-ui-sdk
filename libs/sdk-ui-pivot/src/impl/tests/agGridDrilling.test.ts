// (C) 2007-2020 GoodData Corporation

import { IMappingHeader, createIntlMock, getDrillIntersection } from "@gooddata/sdk-ui";
import { assignDrillItemsAndType, getDrillRowData, getMeasureDrillItem } from "../agGridDrilling";
import { IGridHeader } from "../agGridTypes";
import { getTreeLeaves } from "../agGridUtils";
import { IDimensionItemDescriptor, IResultMeasureHeader } from "@gooddata/sdk-backend-spi";
import { createTableHeaders } from "../agGridHeaders";
import { createRowData } from "../agGridData";
import { DataViewFirstPage, recordedDataView } from "@gooddata/sdk-backend-mockingbird";
import { ReferenceRecordings } from "@gooddata/reference-workspace";

const pivotTableWithColumnAndRowAttributes = recordedDataView(
    ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
    DataViewFirstPage,
);
const intl = createIntlMock();

// TODO: enhance these tests to be parameterized and exercise code with different fixtures (recordings)

describe("getMeasureDrillItem", () => {
    it("should return measure drill item based on response headers", () => {
        const responseHeaders: IDimensionItemDescriptor[] = pivotTableWithColumnAndRowAttributes.dimensionItemDescriptors(
            1,
        );
        const header: IResultMeasureHeader = {
            measureHeaderItem: {
                name: "not important",
                order: 0,
            },
        };

        expect(getMeasureDrillItem(responseHeaders, header)).toMatchSnapshot();
    });
    it("should return null if the header cannot be found", () => {
        const responseHeaders1: IDimensionItemDescriptor[] = pivotTableWithColumnAndRowAttributes.dimensionItemDescriptors(
            0,
        );
        const responseHeaders2: IDimensionItemDescriptor[] = pivotTableWithColumnAndRowAttributes.dimensionItemDescriptors(
            1,
        );
        const header: IResultMeasureHeader = {
            measureHeaderItem: {
                name: "not important",
                order: 99,
            },
        };

        expect(getMeasureDrillItem(responseHeaders1, header)).toBe(null);
        expect(getMeasureDrillItem(responseHeaders2, header)).toBe(null);
    });
});

describe("assignDrillItemsAndType", () => {
    it("should assign measure header item drillItems and type to header", () => {
        const header: IGridHeader = {
            headerName: "test",
            drillItems: [],
        };
        const currentHeader = pivotTableWithColumnAndRowAttributes.allHeaders()[1][1][0];
        const responseHeaders = pivotTableWithColumnAndRowAttributes.dimensionItemDescriptors(1);
        const headerIndex = 0;
        const drillItems: IMappingHeader[] = [];
        assignDrillItemsAndType(header, currentHeader, responseHeaders, headerIndex, drillItems);
        expect(header.type).toEqual("MEASURE_COLUMN");
        expect(header.drillItems).toMatchSnapshot();
    });
    it("should assign attribute header type to header and attribute and attribute value to drillItems", () => {
        const header: IGridHeader = {
            headerName: "test",
            drillItems: [],
        };
        const currentHeader = pivotTableWithColumnAndRowAttributes.allHeaders()[0][0][0];
        const responseHeaders = pivotTableWithColumnAndRowAttributes.dimensionItemDescriptors(0);
        const headerIndex = 0;
        const drillItems: IMappingHeader[] = [];
        assignDrillItemsAndType(header, currentHeader, responseHeaders, headerIndex, drillItems);
        expect(header.type).toEqual("COLUMN_ATTRIBUTE_COLUMN");
        expect(drillItems).toMatchSnapshot();
        // assign empty array to drillItems header. Only leaf headers (measures) should have assigned drill items
        expect(header.drillItems).toEqual([]);
    });
});

describe("getDrillIntersection", () => {
    const tableHeaders = createTableHeaders(pivotTableWithColumnAndRowAttributes.dataView);
    const { rowData } = createRowData(tableHeaders, pivotTableWithColumnAndRowAttributes, intl);
    const columnDefs = tableHeaders.allHeaders;

    it("should return intersection of row attribute and row attribute value for row header cell", async () => {
        const rowColDef = columnDefs[0]; // row header
        const drillItems = [rowData[0].headerItemMap[rowColDef.field], ...rowColDef.drillItems];
        const intersection = getDrillIntersection(drillItems);
        expect(intersection).toMatchSnapshot();
    });

    it("should return intersection of all column header attributes and values and a measure for column header cell", async () => {
        const colDef = getTreeLeaves(columnDefs)[2]; // column leaf header
        const intersection = getDrillIntersection(colDef.drillItems);
        expect(intersection).toMatchSnapshot();
    });

    // tslint:disable-next-line:max-line-length
    it("should return intersection without header property when measure has neither uri nor identifier (arithmetic measure)", async () => {
        const drillItems: IMappingHeader[] = [
            {
                measureHeaderItem: {
                    localIdentifier: "am1",
                    name: "Arithmetic measure",
                    format: "",
                },
            },
        ];
        const intersection = getDrillIntersection(drillItems);
        expect(intersection).toEqual([
            {
                header: {
                    measureHeaderItem: {
                        format: "",
                        localIdentifier: "am1",
                        name: "Arithmetic measure",
                    },
                },
            },
        ]);
    });
});

describe("getDrillRowData", () => {
    const tableHeaders = createTableHeaders(pivotTableWithColumnAndRowAttributes.dataView);
    const { rowData } = createRowData(tableHeaders, pivotTableWithColumnAndRowAttributes, intl);
    const columnDefs = tableHeaders.allHeaders;

    it("should return an array of row data", async () => {
        const leafColumnDefs = getTreeLeaves(columnDefs);
        const drillRow = getDrillRowData(leafColumnDefs, rowData[0]);
        expect(drillRow).toMatchSnapshot();
    });
});
