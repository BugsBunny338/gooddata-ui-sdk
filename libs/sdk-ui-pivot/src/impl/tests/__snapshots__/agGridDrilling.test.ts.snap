// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignDrillItemsAndType should assign attribute header type to header and attribute and attribute value to drillItems 1`] = `
Array [
  Object {
    "attributeHeaderItem": Object {
      "name": "CompuSci",
      "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1054/elements?id=165678",
    },
  },
  Object {
    "attributeHeader": Object {
      "formOf": Object {
        "identifier": "attr.product.id",
        "name": "Product",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1054",
      },
      "identifier": "label.product.id.name",
      "localIdentifier": "a_label.product.id.name",
      "name": "Product Name",
      "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1055",
    },
  },
]
`;

exports[`assignDrillItemsAndType should assign measure header item drillItems and type to header 1`] = `
Array [
  Object {
    "measureHeaderItem": Object {
      "format": "$#,##0.00",
      "identifier": "aangOxLSeztu",
      "localIdentifier": "m_aangOxLSeztu",
      "name": "Amount",
      "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1271",
    },
  },
]
`;

exports[`getDrillIntersection should return intersection of all column header attributes and values and a measure for column header cell 1`] = `
Array [
  Object {
    "header": Object {
      "attributeHeader": Object {
        "formOf": Object {
          "identifier": "attr.owner.region",
          "name": "Region",
          "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1086",
        },
        "identifier": "label.owner.region",
        "localIdentifier": "a_label.owner.region",
        "name": "Region",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1087",
      },
      "attributeHeaderItem": Object {
        "name": "West Coast",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1086/elements?id=460489",
      },
    },
  },
  Object {
    "header": Object {
      "measureHeaderItem": Object {
        "format": "$#,##0.00",
        "identifier": "aangOxLSeztu",
        "localIdentifier": "m_aangOxLSeztu",
        "name": "Amount",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1271",
      },
    },
  },
]
`;

exports[`getDrillIntersection should return intersection of row attribute and row attribute value for row header cell 1`] = `
Array [
  Object {
    "header": Object {
      "attributeHeader": Object {
        "formOf": Object {
          "identifier": "attr.product.id",
          "name": "Product",
          "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1054",
        },
        "identifier": "label.product.id.name",
        "localIdentifier": "a_label.product.id.name",
        "name": "Product Name",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1055",
      },
      "attributeHeaderItem": Object {
        "name": "CompuSci",
        "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1054/elements?id=165678",
      },
    },
  },
]
`;

exports[`getDrillRowData should return an array of row data 1`] = `
Array [
  Object {
    "id": "165678",
    "name": "CompuSci",
  },
  "5099309.54",
  "22123590.1",
]
`;

exports[`getMeasureDrillItem should return measure drill item based on response headers 1`] = `
Object {
  "measureHeaderItem": Object {
    "format": "$#,##0.00",
    "identifier": "aangOxLSeztu",
    "localIdentifier": "m_aangOxLSeztu",
    "name": "Amount",
    "uri": "/gdc/md/l32xdyl4bjuzgf9kkqr2avl55gtuyjwf/obj/1271",
  },
}
`;
